Docs/documentation directory change

kuraisle opened the pull request:
I've moved the docs to a folder github pages can read. If this can be merged quickly, we should be able to serve GH pages ASAP

kuraisle committed:
Added code for vector search on concepts

The main change I've made is adding embeddings.py in components. This provides a class, Embeddings, which can be called from the app to search a vector database of concepts. If the database doesn't exist, it'll build it. This takes a long time, so the database is saved for subsequent runs.

There's also the possibility of running several different embedding databases if necessary.

I need to test whether the search function actually works, as it's currently building the database, taking about an hour.

I've added an API endpoint for the vector search. It uses the same PipelineRequest object to set the options, so that if we mix it with the LLM pipeline we don't have to change anything.

Added the requirements to the environment.yml, I'll update requirements next.

OMOP_match wasn't using the Concept and ConceptSynonym classes directly, so taken those out

Put anything ending with ".qdrant" in the .gitignore, that way as long as we name our databases with this extension we won't have a problem pushing it to GitHub, which would have an issue with 100s MB being uploaded. We might need to find a way of distributing the embedded store

kuraisle committed:
Fixed vector query code

The model needs to be warmed up before running.
I also made the output more legible by just taking the concept name, id, and the score of the match

kuraisle committed:
Updated to match main and multiple input

I hadn't pulled from main, so it was out of date. Hopefully I've caught everything.

Embeddings now takes a list of input

kuraisle committed:
Docs

Documented embeddings, small other docs updates

kuraisle committed:
Merge branch 'main' into feature/embeddings-generator

kuraisle committed:
moved docs to /docs

For compatibility with GitHub Pages

Grazziela commented on Carrot-Assistant/app.py:
There is a typo in names of medications

Grazziela commented on Carrot-Assistant/app.py:
```diff
@@ -227,8 +227,8 @@ async def run_pipeline(request: PipelineRequest) -> EventSourceResponse:
-    request: InformalNameRequest
-        The request containing the informal name of the medication
+    request: PipelineRequest
+        The request containing a list of informal namse of medications
```
typo here in names of medication

Grazziela commented on Carrot-Assistant/app.py:
```diff
@@ -248,7 +248,7 @@ async def run_db(request: PipelineRequest) -> List[Dict[str,Any]]:
     ----------
     request: PipelineRequest
-        An API request containing a medication name
+        An API request containing a list of informal names for medications and the options of a pipeline
```
Could this comment be generic to other informal names, rather than just medication, or would we need to perform a larger refactoring of the tool for that?

kuraisle replied:
At the moment the SQL just runs queries against drug concepts. Extending it to other domains makes the search space much bigger, so a query will be a lot slower (2 minutes/query). There might be other optimizations we could make, but restricting it to drugs was an easy win, bringing queries down to 2-4 seconds.

Grazziela commented on Carrot-Assistant/app.py:
```diff
@@ -271,6 +271,22 @@ async def run_db(request: PipelineRequest) -> List[Dict[str,Any]]:
+    Search a vector database for a medication name
+
+    Default options can be overridden by pipeline_options
+    A warning: if you don't have a vector database set up under the embeddings_path, it will build one. This takes a while, an hour on a 2019 macbook pro.
```
instead of it will build one, shall we say: this method (or this module will build one for you?) instead of saying macbook pro, shall we just add the cpu/memory specs here?

Grazziela commented on Carrot-Assistant/app.py:
```diff
@@ -271,6 +271,22 @@ async def run_db(request: PipelineRequest) -> List[Dict[str,Any]]:
+    Parameters
+    ----------
+    request: PipelineRequest
+        An API request containing a list of informal names for medications and the options of a pipeline
```
again here do we need to specify medications or can it be generic?

Grazziela commented on Carrot-Assistant/components/embeddings.py:
```diff
@@ -14,6 +15,9 @@
 
 class EmbeddingModel(Enum):
+    """
+    An Enum for models used to generate concept embeddings
```
it would be useful to say why this enum structure is needed. To me it tends to be a little more useful, especially for new people looking at the code.

Grazziela commented on Carrot-Assistant/components/embeddings.py:
```diff
@@ -30,6 +42,22 @@ def __init__(
+        Parameters
+        ----------
+        embeddings_path: str
+            A path for the embeddings database. If one is not found, it will be built, which takes a long time. This is built from concepts fetched from the OMOP database
```
shall we point here to where one can find the OMOP database? I wonder if it would be good to explain what is meant by path here. Path to me is just a pointer to an address where the file is located, right? So why would this take long to run? perhaps it is not the path that will be built, but the database itself? Just maybe clarify?

Grazziela commented on Carrot-Assistant/components/embeddings.py:
```diff
@@ -30,6 +42,22 @@ def __init__(
+            A path for the embeddings database. If one is not found, it will be built, which takes a long time. This is built from concepts fetched from the OMOP database
+        force_rebuild: bool
+            If true, the embeddings database will be rebuilt.
+        embed_vocab: List[str]
```
can we change the name of this variable from embed_vocab to embed_vocabulary? I know Reza tends to cut the name of the variables, but I generally prefer longer names that facilitate code reading

Grazziela commented on Carrot-Assistant/components/embeddings.py:
```diff
@@ -30,6 +42,22 @@ def __init__(
+        embed_vocab: List[str]
+            A list of OMOP vocabulary_ids. If the embeddings database is built, these will be the vocabularies used in the OMOP query
+        model: EmbeddingModel
+            The model used to create embeddings
```
can we expand this here? which model could this be? I know that there is a list, but we could hint about the options?

Grazziela commented on Carrot-Assistant/components/embeddings.py:
```diff
@@ -81,13 +112,29 @@ def _build_embeddings(self):
 
     def search(self, query: List[str]) -> List[List[Dict[str,Any]]]:
+        """
+        Search the attached vector database with a list of informal medications
```
informal names rather than medication, if we can generalise?

Grazziela commented on Carrot-Assistant/components/embeddings.py:
```diff
@@ -81,13 +112,29 @@ def _build_embeddings(self):
+        Returns
+        -------
+        List[List[Dict[str, Any]]]
+            For each medication in the query, the result of searching the vector database
```
same here

Grazziela commented on llettuce-docs/_build/html/_sources/autoapi/OMOP_match/index.rst.txt:
```diff
@@ -1,40 +1,41 @@
 
-   This class retrieves matches from an OMOP database and returns the best
 
+   This class retrieves matches from an OMOP database and returns the best
```
based on what matching criteria?

Grazziela commented on llettuce-docs/_build/html/_sources/autoapi/pipeline/index.rst.txt:
```diff
@@ -1,24 +1,26 @@
 
-   This class is used to generate a pipeline for the model
 
+   This class is used to generate a pipeline for the model
```
shall we say what the pipeline stages are here?

Grazziela commented on llettuce-docs/_build/html/_sources/autoapi/prompt/index.rst.txt:
```diff
@@ -1,24 +1,26 @@
 
-   This class is used to generate prompts for the models.
 
+   This class is used to generate prompts for the models.
```
what is the objective of the prompts and how are they generated? worth it adding a few more words here?

Grazziela commented on llettuce-docs/autoapi/OMOP_match/index.rst:
```diff
@@ -1,40 +1,41 @@
 
-   This class retrieves matches from an OMOP database and returns the best
 
+   This class retrieves matches from an OMOP database and returns the best
```
according to which criteria

Grazziela commented on llettuce-docs/autoapi/app/index.rst:
```diff
@@ -202,18 +240,31 @@ Module Contents
 .. py:function:: generate_events(request: PipelineRequest) -> collections.abc.AsyncGenerator[str]
    :async:
 
+   Generate LLM output and OMOP results for a list of informal medication names
```
informal names rather than informal medication names?

Grazziela commented on llettuce-docs/autoapi/app/index.rst:
```diff
@@ -202,18 +240,31 @@ Module Contents
+   Parameters
+   ----------
+   request: PipelineRequest
+       The request containing the list of informal names of the medications.
```
same here

Grazziela commented on llettuce-docs/autoapi/app/index.rst:
```diff
@@ -202,18 +240,31 @@ Module Contents
+   Workflow
+   --------
+   For each informal name:
+       The first event is to Query the OMOP database for a match
```
this is a bit unclear to me. im not sure when the result is just a fetch from the database or when the LLM is being triggered here.

Grazziela commented on llettuce-docs/autoapi/app/index.rst:
```diff
@@ -202,18 +240,31 @@ Module Contents
+   --------
+   For each informal name:
+       The first event is to Query the OMOP database for a match
+       The second event is to fetches relevant concepts from the OMOP database
```
to fetch?

Grazziela commented on llettuce-docs/autoapi/app/index.rst:
```diff
@@ -202,18 +240,31 @@ Module Contents
+   For each informal name:
+       The first event is to Query the OMOP database for a match
+       The second event is to fetches relevant concepts from the OMOP database
+       Finally,The function yields results as they become available,
```
the function, instead of The function? As they become available from the database query or from the LLM matching?

Grazziela commented on llettuce-docs/autoapi/app/index.rst:
```diff
@@ -202,18 +240,31 @@ Module Contents
+   If the OMOP database returns a match, the LLM is not queried
+
+   If the OMOP database does not return a match, 
+   the LLM is used to find the formal name and the OMOP database is 
```
this is unclear to me

Grazziela commented on llettuce-docs/autoapi/app/index.rst:
```diff
@@ -223,40 +274,56 @@ Module Contents
    ----------
    request: PipelineRequest
-       An API request containing a medication name
+       An API request containing a list of informal names for medications and the options of a pipeline
```
if suitable, remove medications and just keep informal names?

Grazziela commented on llettuce-docs/autoapi/app/index.rst:
```diff
@@ -223,40 +274,56 @@ Module Contents
+.. py:function:: run_vector_search(request: PipelineRequest)
+   :async:
+
+   Search a vector database for a medication name
```
same here

Grazziela commented on llettuce-docs/autoapi/app/index.rst:
```diff
@@ -223,40 +274,56 @@ Module Contents
+   Search a vector database for a medication name
+
+   Default options can be overridden by pipeline_options
+   A warning: if you don't have a vector database set up under the embeddings_path, it will build one. This takes a while, an hour on a 2019 macbook pro.
```
change macbook pro to cpu and memory config

Grazziela commented on llettuce-docs/autoapi/app/index.rst:
```diff
@@ -223,40 +274,56 @@ Module Contents
+   Parameters
+   ----------
+   request: PipelineRequest
+       An API request containing a list of informal names for medications and the options of a pipeline
```
remove medication?

Grazziela commented on llettuce-docs/autoapi/pipeline/index.rst:
```diff
@@ -1,24 +1,26 @@
 
-   This class is used to generate a pipeline for the model
 
+   This class is used to generate a pipeline for the model
```
stages of the pipeline?

Grazziela reviewed (changes requested):
@kuraisle reviews to the PR complete

kuraisle committed:
Removed mentions of medication

Changed the docstrings in the app.py so references to medications are more generic, then rebuilt docs
